File: Assign1/task1.scala
object task1_1{
        def main(args:Array[String]):Unit={
        val conf=new SparkConf().setAppName("spark-test").setMaster("local[*]")
        val sc=new SparkContext(conf)
        val textFile=sc.textFile(args(0).trim())

        val row=textFile.map(row=>parse(row))
        //2ND

        val business_ids=row.map(x=>(x \ "business_id")).distinct().count()
//
        //3rd
        val years=args(3).trim()
        val year=row.filter(x=>(x \ "date").values.toString.contains(years))
//
//    //4th

        val user_ids=row.map(x=>((x \ "user_id").values.toString,1))
        val reduce_users=user_ids.reduceByKey((x,y)=>x+y)
        val sorts=reduce_users.sortBy(x=>(-x._2,x._1))

        //5th
        val excluded=List("(","[",",",".","!","?",":",";","]",")","")

        val stopWords=sc.textFile(args(2).trim())
        val all_words=stopWords.map(x=>x.trim().toLowerCase()).collect().toList
        val c=all_words:::excluded
        val text_rdd=row.map(x=>((x \ "text").values.toString)).flatMap(x=>x.toLowerCase().split(" "))


        val filter1=text_rdd.map(removePunc).filter(x=>x!=null&&x!=" "&&!c.contains(x.toLowerCase()))

        val stop_map=filter1.map(x=>(x.toLowerCase(),1))
        val stop_reduce=stop_map.reduceByKey((x,y)=>x+y)
        val stop_sort=stop_reduce.sortBy(x=>(-x._2,x._1)).map(x=>x._1).take(args(5).trim().toInt)

        val file=new File(args(1).trim())
        val bw=new BufferedWriter(new FileWriter(file))
        val map=Map("A"->row.count(),"B"->year.count(),"C"->business_ids,
        "D"->sorts.map(x=>List(JString(x._1),x._2)).take(args(4).trim().toInt),
        "E"->stop_sort.toList)

        bw.write(Json(DefaultFormats).write(map).replace(":",": ").replace(",",", "))
        bw.close()
        sc.stop()


        }
        def removePunc(key:String):String={
        val excluded="([,.!?:;])"
        var values=key
        for(a<-key){
        if(excluded.contains(a)){
        values=values.replace(a.toString,"")
        }
        }
        values.toLowerCase()
        }

        }

File: Assign1/task1.py
from pyspark import SparkContext
import json
import datetime
import sys


def removePunctuation(key) : 
    excluded = ['(', '[', ',', '.', '!', '?', ':', ';' , ']' ,')','']
    for i in excluded : 
        if i in key : 
            key = key.replace(i,"")
    
    return key

sc1 = SparkContext.getOrCreate()
review_json = sc1.textFile(sys.argv[1]).map(lambda k: json.loads(k)) 

#Year
year = sys.argv[4]
years = review_json.filter(lambda x : int(datetime.datetime.strptime(x['date'], '%Y-%m-%d %H:%M:%S').year) == int(year))


#businesses : 
business_id = review_json.map(lambda x: x['business_id'])

#Top Users
top_users = int(sys.argv[5])

user_ids = review_json.map(lambda x: (x['user_id'],1))
reduced  = user_ids.reduceByKey(lambda x,y : x+y)
sort = reduced.takeOrdered(top_users,lambda x :( -x[1],x[0]))
takeTopM = sort


# 5th 
top_n = int(sys.argv[6])
excluded = ['(', '[', ',', '.', '!', '?', ':', ';' , ']' ,')','']
file1 = open(sys.argv[3],"r")
x  = file1.readline()
while(x != "") : 
    excluded.append(x.lower().strip())
    x = file1.readline()
text = review_json.map(lambda x: x['text'])
map1 = text.flatMap(lambda text: text.lower().split(' '))
filter1 = map1.map(removePunctuation).filter(lambda x : x != None and x!="" and x.lower() not in excluded )
stop_map = filter1.map(lambda x :(x,1))
stop_reduce = stop_map.reduceByKey(lambda x,y : x+y)
stop_sort = stop_reduce.sortBy(lambda x : (-x[1],x[0])).map(lambda x :x[0]).take(top_n)
print(stop_sort)
ans = {
    "A": review_json.count(),
    "B": years.count(),
    "C": business_id.distinct().count(), 
    "D": takeTopM,
    "E": stop_sort,
}
f = open(sys.argv[2],"w")
json.dump(ans, f)
f.close()
sc1.stop()

